### PR TITLE
Make autogen ID loading work with custom type

### LINKED ISSUE #2615: Primary Key ignores custom Ecto.Type#load after Repo.insert, Repo.all since 2.2.0
When updating from 2.1.6 to 2.2.0 (or higher) we are running into troubles with "deserializing" custom Ecto.Types for primary key. It works fine with normal columns. All our primary keys are UUIDs inside the Postgresql DB .

Our Ecto.Type `Custom.PrefixedId` transparently adds/removes a prefix to the ids. So clients can work with friendly uuids that indicate their type: e.g. `blog_12323-..-23121`, `post_12323-..-123123`

```elixir
defmodule Blog.Post do
  # adds a "virtual" prefix to a uuid column.  
  alias Custom.PrefixedId

  @primary_key {:id, PrefixedId, autogenerate: true}
  @derive {Phoenix.Param, key: :id}
  @foreign_key_type Ecto.UUID

  schema "posts" do  
    field(:other_id, PrefixedId) # database type is :uuid
  end
end
```

## BEFORE. Correct behaviour with ecto 2.1.6

```elixir
b = Repo.insert!(%Blog.Post{other_id: UUID.uuid4})
b.id       # => lorem_c0690cf7-3e57-4338-bb0c-9e2279d1d242
b.other_id # => lorem_86399b0f-f9b7-4fd5-93ea-c4665f991cc1

b = Repo.get(Blog.Post, "lorem_c0690cf7-3e57-4338-bb0c-9e2279d1d242")
b.id       # => "lorem_c0690cf7-3e57-4338-bb0c-9e2279d1d242"
b.other_id # => "lorem_86399b0f-f9b7-4fd5-93ea-c4665f991cc1"

[b] = Repo.all(Blog.Post)
b.id       # => "lorem_c0690cf7-3e57-4338-bb0c-9e2279d1d242"
b.other_id # => "lorem_86399b0f-f9b7-4fd5-93ea-c4665f991cc1"
```

## AFTER. with ecto ~> 2.2.0 (all versions)

Blog.Post#id is no longer consistently "deserialized", but other columns with the same Ecto.Type are. 

```elixir
b = Repo.insert!(%Blog.Post{other_id: UUID.uuid4})
b.id       # => c0690cf7-3e57-4338-bb0c-9e2279d1d242
b.other_id # => lorem_86399b0f-f9b7-4fd5-93ea-c4665f991cc1

# But with get it seems to work just fine
b = Repo.get(Blog.Post, "lorem_c0690cf7-3e57-4338-bb0c-9e2279d1d242")
b.id       # => "lorem_c0690cf7-3e57-4338-bb0c-9e2279d1d242" 
b.other_id # => "lorem_86399b0f-f9b7-4fd5-93ea-c4665f991cc1"

# Updating that records other_id keeps the id correct, but does not prefix other_id
b = Repo.get(Blog.Post, "lorem_c0690cf7-3e57-4338-bb0c-9e2279d1d242")
u = Ecto.Changeset.change(b, other_id: UUID.uuid4) |> Repo.update!
u.id       # => "lorem_c0690cf7-3e57-4338-bb0c-9e2279d1d242" 
u.other_id # => "000e235d-e77b-4ef5-91a1-e7bfc3db130e"

[b] = Repo.all(Blog.Post)
b.id       # => "lorem_c0690cf7-3e57-4338-bb0c-9e2279d1d242"
b.other_id # => "lorem_86399b0f-f9b7-4fd5-93ea-c4665f991cc1"
```

Most probably somehow the Ecto.Type given in @primary_key is not properly registered. Or not consistently applied with primary key columns.

### Environment:

```
PostgreSQL 9.6
Erlang/OTP 20 [erts-9.2.1] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:10] [hipe] [kernel-poll:false]
Elixir 1.6.4 (compiled with OTP 19)
```

#### Updates
- I made a mistake, `Repo.all` is actually working fine in the new version. Updated description.
- Updated with a description when updating

### PR DESCRIPTION
**Pull request by qcam**
An attempt to fix #2615.

Besides of fixing the issue, this PR changes/fixes the behavior of `Ecto.Repo.__schema__(:autogenerate_id)` when it comes to custom type ID.

```ex
defmodule MyUUID do
  @behaviour Ecto.Type
  def type(), do: :binary_id
end

defmodule Post do
  use Ecto.Schema

  @primary_key {:id, MyUUID, autogenerate: true}

  schema "posts" do
    # fields eaten by the Internet.
  end
end
```

The current behavior of the API returns `{:id, :id, :binary_id}`. This PR changes it to `{:id, :id, MyUUID}`.

IMHO this behavior is more correct, but we could come up with another fix if this change is not desired.

### CODE REVIEW

File: test/ecto/type_test.exs [hunk 1]
```diff
@@ -776,6 +802,11 @@ defmodule Ecto.TypeTest do
     end
   end
 
+  test "#2615 case" do
```
Comment (qcam) [inline test/ecto/type_test.exs]: This will be updated if the approach is accepted ^__^.

File: test/ecto/type_test.exs [hunk 2]
```diff
@@ -34,6 +45,15 @@ defmodule Ecto.TypeTest do
     end
   end
 
+  defmodule PrefixedSchema do
```
Comment (josevalim) [inline test/ecto/type_test.exs]: Let's call this `PrefixedIDSchema` for clarity?

File: test/ecto/type_test.exs [hunk 3]
```diff
@@ -34,6 +45,15 @@ defmodule Ecto.TypeTest do
...
+    use Ecto.Schema
+
+    @primary_key {:id, PrefixedID, autogenerate: true}
+
```
Comment (josevalim) [inline test/ecto/type_test.exs]: We can remove this new line. :)

**Comment by josevalim** [comment]
@qcam it looks perfect to me. Let's tidy this up and merge it. :heart:

**Comment by josevalim** [comment]
@qcam we also need to change [this line](https://github.com/elixir-ecto/ecto/blob/e1adc8aef7b0f3dbac16620b41eaec03aa790c74/lib/ecto/repo/schema.ex#L117) to call the new function you added in `Ecto.Type`.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by josevalim** [comment]
I will change it locally as I want to use this opportunity and normalize the autogenerate API between schemas and adapters. :)

**Pull request merged**